### PR TITLE
Robust resolution of deeply nested server values.

### LINKED ISSUE #1013: Resolution of deeply nested server values crashes the client
Following #1009 deeply nested server values causes the client to crash, because the resolution algorithm is now recursive.

Here is an example that exhibits the behaviour (the crash is machine dependent, so you may have to increase the length of the list `xs`)
```links
var xs = [0..2085];

fun mainPage(_) {
  page
    <#>
      <button l:onclick="{println(intToString(length(xs)))}">Click me</button>
    </#>
}

fun main() {
  addRoute("/", mainPage);
  servePages()
}

main()
```

The fix is to turn it back into iterative form.

### PR DESCRIPTION
**Pull request by dhil**
This patch converts the function `resolveServerValue` into CPS with a
trampoline such that resolution of deeply nested server values no
longer causes a stack overflow.

Resolves #1013.

### CODE REVIEW

**Comment by dhil** [comment]
@SimonJF do you mind testing this patch with your examples?

File: lib/js/jslib.js [hunk 1]
```diff
@@ -772,6 +772,17 @@ const LINKS = new (function() {
...
+   // Server value resolution
+   class ServerValueResolutionYield extends Error {
+     constructor(k) {
+       super("resolveServerValueCPS yielding");
+       this.k = k;
+     }
+   }
+
+   let _resolveDepth = 0;
+   let _resolveDepthMax = 1000;
```
Comment (SimonJF) [inline lib/js/jslib.js:784]: Minor, but this can be a `const`, right?

Comment (SimonJF) [inline lib/js/jslib.js:784]:
```suggestion
   const _resolveDepthMax = 1000;
```

**Review by SimonJF — APPROVED** [review]
This seems to work on my GtoPdb examples. There is an issue with the session-typed chatserver, but I think that's separate as I tried it before and after this.

**Comment by dhil** [comment]
I will look at your session type example. I have a feeling it is related to value resolution

**Comment by dhil** [comment]
@SimonJF please test again. It should work now.

**Comment by SimonJF** [comment]
Confirmed working. Happy for this to go in at your convenience.

**Comment by SimonJF** [comment]
And thanks for looking at the broken example!

**Pull request merged**